### PR TITLE
Add test driver py copy from vm

### PR DESCRIPTION
**Pull request by 7c6f434c**
###### Motivation for this change

Add functionality for convenient copying from VM to `$out` via the shared dir

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @flokli

### CODE REVIEW

File: nixos/lib/test-driver/test-driver.py [hunk 1]
```diff
@@ -525,6 +527,35 @@ def screenshot(self, filename: str) -> None:
             if ret.returncode != 0:
                 raise Exception("Cannot convert screenshot")
 
+    def copy_from_vm(self, source: str, target_dir: str = "") -> None:
+        """Copy a file from the VM (specified by an in-VM source path) to $out
+        (with an optional subdirectory parameter). The file is copied via the
+        shared directory.
```
Comment (flokli) [inline nixos/lib/test-driver/test-driver.py]: What about

> to a path relative to `$out`

> via the `shared_dir` shared amongst all VMs.

File: nixos/lib/test-driver/test-driver.py [hunk 2]
```diff
@@ -525,6 +527,35 @@ def screenshot(self, filename: str) -> None:
...
+        """
+        # Compute the source, target, and intermediate shared file names
+        out_dir = pathlib.Path(os.environ.get("out", os.getcwd()))
+        src_orig = pathlib.Path(source)
+        int_shared_dir = pathlib.Path("/tmp/xchg")
+        with tempfile.TemporaryDirectory(dir=self.shared_dir) as shared_td:
+            shared_temp = pathlib.Path(shared_td)
+            shared_temp_in = int_shared_dir / shared_temp.name
```
Comment (flokli) [inline nixos/lib/test-driver/test-driver.py]: Can you inline `int_shared_dir` into here?

Comment (flokli) [reply nixos/lib/test-driver/test-driver.py]: I'd prefix paths valid inside the VM context with `vm_`.

File: nixos/lib/test-driver/test-driver.py [hunk 3]
```diff
@@ -525,6 +527,35 @@ def screenshot(self, filename: str) -> None:
...
+            src_in = shared_temp_in / src_orig.name
+            src_out = shared_temp / src_orig.name
+            tgt = out_dir / target_dir
```
Comment (flokli) [inline nixos/lib/test-driver/test-driver.py]: this is the absolute path of target_dir. I'd suggest to inline that or at least move if immediately before the actual copy action.

**Comment by worldofpeace** [comment]
:sparkles: Was just wanting a method like this.

**Comment by 7c6f434c** [comment]
Added `make_command` function to auto-quote a lot of arguments before passing to VM; not sure if looking for or implementing advanced type-based approach is worth it (with type `Literal` which is not quoted and the result of quoting being `Literal`)

**Comment by 7c6f434c** [comment]
@flokli Any opinion on my updated version?

**Comment by flokli** [comment]
Sorry, forgot about that. LGTM.

**Comment by srhb** [comment]
This seems to have broken sharing files between VMs -- at least it looks that way in the ceph tests. Are we supposed to use a different directory now from within the VMs?

**Comment by 7c6f434c** [comment]
Hm, I think I saw it used as an external path to shared directory (which predictably did not work). So it should be split into two paths then?

**Comment by srhb** [comment]
I'm actually not sure. It feels like there should be one shared directory on the host that all VMs are able to access. I don't think we have to worry about compatibility, so whether we make this directory distinct from the one used by copy probably doesn't matter.

**Comment by 7c6f434c** [comment]
Well, when I was writing the copy code — I cannot exclude something has changed since then — I saw that `/tmp/shared` is not the (external) path used for the shared directory. I guess the problem is with using the variable with the external path as internal path?

**Comment by srhb** [comment]
Not sure I understood that, but, yes it's a problem if the variable now points to the vm-specific dirs on the host-side, whereas previously it was pointing into one directory on the host side for all VMs.

**Comment by 7c6f434c** [comment]
When I was changing it, the variable was pointing to a directory that did not exist during the test (host-side)

**Pull request merged**